lib/streamlit/runtime.py

```diff
@@ -0,0 +1,591 @@
+        """Initialize a SessionInfo instance.
+
+        Parameters
+        ----------
+        session : AppSession
+            The AppSession object.
+        client : SessionClient
+            The concrete SessionClient for this session.
+        """
+        self.session = session
+        self.client = client
+        self.script_run_count = 0
+
+
+class RuntimeState(Enum):
+    INITIAL = "INITIAL"
+    WAITING_FOR_FIRST_SESSION = "WAITING_FOR_FIRST_SESSION"
+    ONE_OR_MORE_SESSIONS_CONNECTED = "ONE_OR_MORE_SESSIONS_CONNECTED"
+    NO_SESSIONS_CONNECTED = "NO_SESSIONS_CONNECTED"
+    STOPPING = "STOPPING"
+    STOPPED = "STOPPED"
+
+
+class Runtime:
+    def __init__(self, config: RuntimeConfig):
+        """Create a StreamlitRuntime. It won't be started yet.
+
+        StreamlitRuntime is *not* thread-safe. Its public methods are
+        generally safe to call only on the same thread that its event loop runs
+        on.
+
+        Parameters
+        ----------
+        config
+            Config options.
+        """
+        # Will be set when we start.
+        self._eventloop: Optional[asyncio.AbstractEventLoop] = None
+
+        self._script_path = config.script_path
+        self._command_line = config.command_line or ""
+
+        # Mapping of AppSession.id -> SessionInfo.
+        self._session_info_by_id: Dict[str, SessionInfo] = {}
+
+        self._state = RuntimeState.INITIAL
+
+        # Set after Runtime.stop() is called. Never cleared.
+        self._must_stop = asyncio.Event()
+
+        # Set when a client connects; cleared when we have no connected clients.
+        self._has_connection = asyncio.Event()
+
+        # Set after a ForwardMsg is enqueued; cleared when we flush ForwardMsgs.
+        self._need_send_data = asyncio.Event()
+
+        # Initialize managers
+        self._message_cache = ForwardMsgCache()
+        self._uploaded_file_mgr = UploadedFileManager()
+        self._uploaded_file_mgr.on_files_updated.connect(self._on_files_updated)
+
+        self._stats_mgr = StatsManager()
+        self._stats_mgr.register_provider(get_memo_stats_provider())
+        self._stats_mgr.register_provider(get_singleton_stats_provider())
+        self._stats_mgr.register_provider(_mem_caches)
+        self._stats_mgr.register_provider(self._message_cache)
+        self._stats_mgr.register_provider(in_memory_file_manager)
+        self._stats_mgr.register_provider(self._uploaded_file_mgr)
+        self._stats_mgr.register_provider(
+            SessionStateStatProvider(self._session_info_by_id)
+        )
+
+    @property
+    def state(self) -> RuntimeState:
+        return self._state
+
+    @property
+    def message_cache(self) -> ForwardMsgCache:
+        return self._message_cache
+
+    @property
+    def uploaded_file_mgr(self) -> UploadedFileManager:
+        return self._uploaded_file_mgr
+
+    @property
+    def stats_mgr(self) -> StatsManager:
+        return self._stats_mgr
```

Comment:
We're currently just exposing our singleton manager instances so that Server can use them. This might change, depending on how we do manager-injection? (It depends on whether these managers become interfaces with custom implementations.)